keps/sig-api-machinery/20191001-oasv3.md

```diff
@@ -0,0 +1,938 @@
+---
+title: Kubernetes's OpenAPI v3 Specification
+authors:
+  - "@p0lyn0mial"
+owning-sig: sig-api-machinery
+participating-sigs:
+  - sig-api-machinery
+reviewers:
+  - "@sttts"
+  - @yue9944882
+approvers:
+  - "@sttts"
+  - @yue9944882
+editor: TBD
+creation-date: 2019-10-01
+last-updated: 2019-10-01
+status: provisional
+---
+
+# Kubernetes's OpenAPI v3 Specification
+
+## Table of Contents
+
+TBD
+
+## Release Signoff Checklist
+
+**ACTION REQUIRED:** In order to merge code into a release, there must be an issue in [kubernetes/enhancements] referencing this KEP and targeting a release milestone **before [Enhancement Freeze](https://github.com/kubernetes/sig-release/tree/master/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core Kubernetes i.e., [kubernetes/kubernetes], we require the following Release Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+TDB
+
+## Motivation
+
+TBD
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+
+### Goals
+- Define OpenAPI v3 Specification that will be published by the API servers
+- Improve the names of the object
+- Reuse parameters
+- Publish specifications per API group instead of one document for all groups
```

Comment:
the goal is:
- reduce CPU load due to spec merging
- align aggregated API server OpenAPI specs with per-group APIServices
- allow aggregated API servers to publish v2 and v3 to kube-aggregator, without doubling the CPU load